### DISCUSSION TITLE
How to mock class that uses office-js?

### DISCUSSION THREAD

**Discussion by rosergeev** [vitest-dev/vitest#5305] [Q&A] [answered]
We are migrating our tests from jest to vitest and have encountered with the problem that don't know how to solve. There is a component, let's say ButtonWrapper that onClick event calls a method getHost which in turn uses some code from office-js library(e.g. Office.context.host). Build of the project is successful and doesn't cause any issues. But in tests I'd like to mock module that uses Office namespace, but always got an error that "Office is not defined"

Here it is minimal reproduction for this:
https://codesandbox.io/p/sandbox/testofficelib-v9q2zl
1. Component resides under src/components/ButtonWrapper.tsc
2. Class with call to office-js api - src/api/officeApi.ts
3. Test is located here - src/tests/ButtonWrapper.test.tsx

How do mock modules that use office-js library correctly?

**Comment by hi-ogawa** [comment, marked as answer]
> Build of the project is successful and doesn't cause any issues

In your reproduction, I don't see `@microsoft/office-js` is used anywhere. Can you also show how you're integrating this library into your actual app?

Since we aren't necessary familiar with this library, showing how it's intended to be used within the app would give hints about how to mock it during the tests.

**Reply by hi-ogawa** [reply]
Ah, I think I misunderstood what you're trying to do. You are not trying to mock `@microsoft/office-js` but trying to mock your wrapper `src/api/officeApi.ts` which uses `@microsoft/office-js`.

One way to do that is to mock entire `OfficeApi` named export via `vi.mock`:

https://stackblitz.com/edit/github-cwo1l5?file=src%2Ftests%2FButtonWrapper.test.tsx

```ts
vi.mock('./../api/officeApi', () => {
  class OfficeApi {
    getHost() {
      return 'SomeHost';
    }
  }
  return { OfficeApi };
});
```

This might not be what you're expecting, but the idea is that you need to `return` named exports as an object.

The other way I can think of is to straight-up mutate `OfficeApi.prototype` to change `getHost` without `vi.mock`. Something like this:

https://stackblitz.com/edit/github-cwo1l5-hh9vfu?file=src%2Ftests%2FButtonWrapper.test.tsx

```ts
vi.spyOn(OfficeApi.prototype, "getHost").mockReturnValue("SomeHost");
```

If you are migrating from Jest, I think it would be nice if you can show how you were using `jest.mock` before. With that, it's easier to provide an equivalent alternative on Vitest.

**Reply by rosergeev** [reply]
office-js library is used to develop Add-Ins for Microsoft Office Applications(Word, Excel, Outlook etc). Office App is considered to be a host application and Add-In (html + javascript, could be angular/react/vue or whatever) is loaded to a separate sandbox, actually a new Frame. office-js library works as a proxy between host app and Add-In, so our code can communicate with host **only** through office-js api calls. When office-js is loaded it verifies that it was loaded in context of the Office App, otherwise it logs to console an error or warning that office-js is loaded outside the Office Application context. It would be a bit complicated to introduce an example of usage of this library.
You can take a look at quick overview of the Office Add-In Platform here:
https://learn.microsoft.com/en-us/office/dev/add-ins/overview/office-add-ins

**Reply by rosergeev** [reply]
Am I right thinking that in first approach we completely replace OfficeApi module with new implementation and that´s how we eliminate any mention of Office namespace? I do not understand why doesn´t work the following:
```
vi.mock("./../api/officeApi", async () => {
  const actual = await vi.importActual("./../api/officeApi");

  return {
    ...actual,
    getHost: vi.fn().mockReturnValue("SomeHost"),
  };
});
```
Here I use named export as return as well, but this cause an error. I guess due to while importing actual we are encountering with the original code that has mention of Office namespace. 
Also I don't know is it possible to mock some of the module functions and then expect on them? E.g. I'd like to do the following:
```
const mockedFn = vi.fn()
vi.mock("someModule", async () =>{
  const mod = await vi.importActual("someModule")
  return {
    ...mod,
    someMethod: mockedFn
  }
})
...
test("Test of the component", () =>{
...
expect(mockedFn).toBeCalled()
})
```
